snitun/server/worker.py

```diff
@@ -0,0 +1,116 @@
+"""SniTun worker for traffics."""
+import asyncio
+from multiprocessing import Process, Manager, Queue
+from threading import Thread, Event
+from typing import Dict, Optional, List, Tuple
+from socket import socket
+
+from .listener_peer import PeerListener
+from .listener_sni import SNIProxy
+from .peer_manager import PeerManager, PeerManagerEvent
+from .peer import Peer
+
+
+class ServerWorker(Process):
+    """Worker for multiplexer."""
+
+    def __init__(
+        self,
+        fernet_keys: List[str],
+        throttling: Optional[int] = None,
+    ) -> None:
+        """Initialize worker & communication."""
+        super().__init__()
+
+        self._fernet_keys: List[str] = fernet_keys
+        self._throttling: Optional[int] = throttling
+
+        # Used on the child
+        self._peers: Optional[PeerManager] = None
+        self._list_sni: Optional[SNIProxy] = None
+        self._list_peer: Optional[PeerListener] = None
+        self._loop: Optional[asyncio.BaseEventLoop] = None
+
+        # Communication between Parent/Child
+        self._manager: Manager = Manager()
+        self._new: Queue = self._manager.Queue()
+        self._sync: Dict[str, None] = self._manager.dict()
+
+    def is_responsible_peer(self, sni: str) -> bool:
+        """Return True if worker is responsible for this peer domain."""
+        return sni in self._sync
+
+    async def _async_init(self) -> None:
+        """Initialize child process data."""
+        self._peers = PeerManager(
+            self._fernet_keys,
+            throttling=self._throttling,
+            event_callback=self._event_stream,
+        )
+        self._list_sni = SNIProxy(self._peers)
+        self._list_peer = PeerListener(self._peers)
+
+    def _event_stream(self, peer: Peer, event: PeerManagerEvent) -> None:
+        """Event stream peer connection data."""
+        if event == PeerManagerEvent.CONNECTED:
+            self._sync[peer.hostname] = None
+        else:
+            self._sync.pop(peer.hostname, None)
+
+    def shutdown(self) -> None:
+        """Shutdown child process.
+
+        This function blocking, don't call it inside loop!
+        """
+        self._new.put(None)
+        self.join(10)
+
+    def handover_connection(
+        self, con: socket, data: bytes, sni: Optional[str] = None
+    ) -> None:
+        """Move new connection to worker.
+
+        Async friendly.
+        """
+        self._new.put_nowait((con, data, sni))
+
+    def run(self) -> None:
+        """Running worker process."""
+        self._loop = asyncio.get_event_loop()
```

Comment:
Is it safe to use `asyncio.get_event_loop` when we use another thread to run the loop below? `asyncio.get_event_loop` will set the loop to the current OS thread, right?

Reply:
Yes, it's just the loop object and since we are on the fork, we get the forked loop from the parent process but it is only initialized and also not running. Of curse we can create a new event loop, but why we should not reuse the available one which is maybe set up to use uvloop.

Reply:
Right, if the event loop uses epoll, they share it over the process. I change it now because we do not know what kind of selectors are used